src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs

```diff
@@ -0,0 +1,995 @@
+
+            // *** Binary format ***
+            // int: number of added columns
+            // for each added column
+            //   string: output column name
+            //   string: input column name
+            // Binary stream of mapping
+
+            var length = ctx.Reader.ReadInt32();
+            var columns = new (string Source, string Name)[length];
+            for (int i = 0; i < length; i++)
+            {
+                columns[i].Name = ctx.LoadNonEmptyString();
+                columns[i].Source = ctx.LoadNonEmptyString();
+            }
+
+            byte[] rgb = null;
+            Action<BinaryReader> fn = r => rgb = ReadAllBytes(env, r);
+
+            if (!ctx.TryLoadBinaryStream(DefaultMapName, fn))
+                throw env.ExceptDecode();
+
+            var binaryLoader = GetLoader(env, rgb);
+            var keyColumnName = (termLookupModel) ? "Term" : KeyColumnName;
+            return new ValueMappingTransformer(env, binaryLoader, keyColumnName, ValueColumnName, columns);
+        }
+
+        private static byte[] ReadAllBytes(IExceptionContext ectx, BinaryReader rdr)
+        {
+            Contracts.AssertValue(ectx);
+            ectx.AssertValue(rdr);
+            ectx.Assert(rdr.BaseStream.CanSeek);
+
+            long size = rdr.BaseStream.Length;
+            ectx.CheckDecode(size <= int.MaxValue);
+
+            var rgb = new byte[(int)size];
+            int cb = rdr.Read(rgb, 0, rgb.Length);
+            ectx.CheckDecode(cb == rgb.Length);
+
+            return rgb;
+        }
+
+        protected static IDataTransform Create(IHostEnvironment env, ModelLoadContext ctx, IDataView input)
+            => Create(env, ctx).MakeDataTransform(input);
+
+        private static IRowMapper Create(IHostEnvironment env, ModelLoadContext ctx, Schema inputSchema)
+            => Create(env, ctx).MakeRowMapper(inputSchema);
+
+        protected static PrimitiveType GetPrimitiveType(Type rawType, out bool isVectorType)
+        {
+            Type type = rawType;
+            isVectorType = false;
+            if (type.IsArray)
+            {
+                type = rawType.GetElementType();
+                isVectorType = true;
+            }
+
+            if  (!type.TryGetDataKind(out DataKind kind))
+            {
+                throw new InvalidOperationException($"Unsupported type {type} used in mapping.");
+            }
+
+            return PrimitiveType.FromKind(kind);
+        }
+
+        public override void Save(ModelSaveContext ctx)
+        {
+            Host.CheckValue(ctx, nameof(ctx));
+            ctx.SetVersionInfo(GetVersionInfo());
+            SaveColumns(ctx);
+
+            // Save out the byte stream of the IDataView of the data source
+            ctx.SaveBinaryStream(DefaultMapName, w => w.Write(_dataView));
+        }
+
+        /// <summary>
+        /// Base class that contains the mapping of keys to values.
+        /// </summary>
+        private abstract class ValueMap
```

Comment:
>ValueMap [](start = 31, length = 8)

You have only one child class for this abstract class, is it worth to have it?

Reply:
The main issue with combining the two classes and having a single class is the declaration of the ValueMap on the Transformer. It would go from ValueMap _valueMap to ValueMap<TKey, TValue> _valueMap -- this would be OK if TKey and TValue could be used directly, however the ValueMap is constructed based on the column types (which are similar!). Where they differr is when TValue is a vector type. When creating ValueMapping Transformer with an int vector, TValue is still only in. However in the value map its int[].

---
In reply to: [243464824](https://github.com/dotnet/machinelearning/pull/1710#discussion_r243464824) [](ancestors = 243464824)